rfcs/20240722-group-query-attention/README.md

```diff
@@ -0,0 +1,222 @@
+# Support Grouped-query Attention in Graph API
+
+## Introduction & Motivation
+
+In typical Multi-Head Attention (MHA), Query, Key, and Value tensors usually
+have the same head number (eg., all Query, Key, and Value have shape (N, H, S,
+D) where N is mini-batch size, S is sequence length, H is head number, and D is
+head size). It becomes the performance bottleneck to load the Key and Value
+tensors in each generation step especially when the sentence length becomes
+longer.
+
+To reduce the memory bandwidth of loading the Key and Value tensors, Multi-Query
+Attention (MQA) is invented by reducing the head number of Key and Value to 1
+which means multiple Queries will map the same single Key and Value (the shape
+of Key and Value become (N, 1, S, D)).
+
+However, MQA can lead to model quality degradation and training instability.
+Grouped-Query Attention (GQA), an interpolation between typical MHA and MQA, is
+proposed with single Key and Value head per subgroup of Query heads (the shape
+of Key and Value become (N, G, S, D) where H > G and H % G = 0). GQA is widely
+adopted in llama2-70B, llama3 family, llama3.1 family [[#1]][1], Mistral, and
+StarCoder2 [[#2]][2].
+
+The following figure is from the paper [[#3]][3] and shows the comparison of
+MHA, MQA, and GQA.
+
+![attention](attention.png)
+
+oneDNN already supports Scaled Dot-Product Attention (SDPA) through Graph API
+([document](https://oneapi-src.github.io/oneDNN/dev_guide_graph_sdpa.html#doxid-dev-guide-graph-sdpa)
+and
+[examples](https://github.com/oneapi-src/oneDNN/blob/main/examples/graph/sdpa.cpp)).
+This proposal aims to extend the support from SDPA to cover GQA.
+
+## GQA in PyTorch
+
+Unlike SDPA, PyTorch does not support GQA as a fused operations. In Huggingface
```

Comment:
FYI - the PyTorch PR just got merged this week: https://github.com/pytorch/pytorch/pull/132689